src/utilities.jl

```diff
@@ -1,5 +1,27 @@
+# Utility function for getting chunks of an ND-array along a particular dimension
+chunks(A, k::Int; dim::Int = 1) = 
+    (selectdim(A, dim, i) for i in Iterators.partition(axes(A,dim), cld(size(A,dim), k)));
```

Comment:
I'm surprised the `gradtest` is passing for `Vit` since Zygote has troubles with this:
```julia
julia> chunks(A, k::Int; dim::Int = 1) = 
           (selectdim(A, dim, i) for i in Iterators.partition(axes(A,dim), cld(size(A,dim), k)));

julia> gradient(x -> sum(chunks(x, 2)[1]), rand(5))
ERROR: MethodError: no method matching getindex(::Base.Generator{Base.Iterators.PartitionIterator{Base.OneTo{Int64}}, var"#13#14"{Int64, Vector{Float64}}}, ::Int64)
Stacktrace:
  [1] macro expansion
    @ ~/.julia/packages/Zygote/FPUm3/src/compiler/interface2.jl:0 [inlined]
  [2] _pullback(::Zygote.Context, ::typeof(getindex), ::Base.Generator{Base.Iterators.PartitionIterator{Base.OneTo{Int64}}, var"#13#14"{Int64, Vector{Float64}}}, ::Int64)
    @ Zygote ~/.julia/packages/Zygote/FPUm3/src/compiler/interface2.jl:9
  [3] _pullback
    @ ~/.julia/packages/Zygote/FPUm3/src/tools/builtins.jl:15 [inlined]
  [4] _pullback(::Zygote.Context, ::typeof(Zygote.literal_getindex), ::Base.Generator{Base.Iterators.PartitionIterator{Base.OneTo{Int64}}, var"#13#14"{Int64, Vector{Float64}}}, ::Val{1})
    @ Zygote ~/.julia/packages/Zygote/FPUm3/src/compiler/interface2.jl:0
  [5] _pullback
    @ ./REPL[8]:1 [inlined]
  [6] _pullback(ctx::Zygote.Context, f::var"#15#16", args::Vector{Float64})
    @ Zygote ~/.julia/packages/Zygote/FPUm3/src/compiler/interface2.jl:0
  [7] _pullback(f::Function, args::Vector{Float64})
    @ Zygote ~/.julia/packages/Zygote/FPUm3/src/compiler/interface.jl:34
  [8] pullback(f::Function, args::Vector{Float64})
    @ Zygote ~/.julia/packages/Zygote/FPUm3/src/compiler/interface.jl:40
  [9] gradient(f::Function, args::Vector{Float64})
    @ Zygote ~/.julia/packages/Zygote/FPUm3/src/compiler/interface.jl:75
 [10] top-level scope
    @ REPL[8]:1
 [11] top-level scope
    @ ~/.julia/packages/CUDA/bki2w/src/initialization.jl:52
```
What's the explanation?

Reply:
The `gradtest` is skipped for all models in Metalhead😅

Reply:
Ah! Ok, we can address that in another PR

Reply:
If it's the time/resources factor for CI being a problem, maybe we could add a message in the Contributions document asking for people to run `gradtest` locally. I'm doing that now just to check if the previous contributions are solid too. And I think the above failure means that this PR will have to wait for https://github.com/JuliaML/MLUtils.jl/pull/47 to be merged and Metalhead will have to take on MLUtils as a dep, right?

Reply:
I don't think CI time would be a problem given that we just check that a single gradient doesn't error out, so we don't do expensive finite-difference comparisons etc... 

> And I think the above failure means that this PR will have to wait for https://github.com/JuliaML/MLUtils.jl/pull/47 to be merged and Metalhead will have to take on MLUtils as a dep, right?

yes we can go that way (and in any case Flux will depend on MLUtils as well, soonish). I just merged the PR and tagged a new MLUtils release, so here we should have a compat MLUtils >= 0.1.1

Reply:
<img width="383" alt="Screenshot 2022-02-10 at 1 14 54 PM" src="https://user-images.githubusercontent.com/74202102/153360735-e5c20fcd-3517-4649-bab8-82d77ec5d1e8.png">
This is on an M1 Mac with 16 GB RAM...and while the tests pass, I am very surprised at the times taken for execution....is this normal? MobileNetv3 and DenseNet are especially surprising

Reply:
> yes we can go that way (and in any case Flux will depend on MLUtils as well, soonish). I just merged the PR and tagged a new MLUtils release, so here we should have a compat MLUtils >= 0.1.1

Will make the change 👍🏽

Reply:
Getting this error when I try to add MLUtils:

```
ERROR: The following 1 direct dependency failed to precompile:

MLUtils [f1d291b0-491e-4a28-83b9-f70985020b54]

Failed to precompile MLUtils [f1d291b0-491e-4a28-83b9-f70985020b54] to /Users/theabhirath/.julia/compiled/v1.8/MLUtils/jl_PwZka8.
ERROR: LoadError: cannot replace module Datasets during compilation
Stacktrace:
 [1] include(mod::Module, _path::String)
   @ Base ./Base.jl:421
 [2] include(x::String)
   @ MLUtils.Datasets ~/.julia/packages/MLUtils/2lblx/src/Datasets/Datasets.jl:1
 [3] top-level scope
   @ ~/.julia/packages/MLUtils/2lblx/src/Datasets/Datasets.jl:5
 [4] include(mod::Module, _path::String)
   @ Base ./Base.jl:421
 [5] include(x::String)
   @ MLUtils ~/.julia/packages/MLUtils/2lblx/src/MLUtils.jl:1
 [6] top-level scope
   @ ~/.julia/packages/MLUtils/2lblx/src/MLUtils.jl:68
 [7] include
   @ ./Base.jl:421 [inlined]
 [8] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
   @ Base ./loading.jl:1399
 [9] top-level scope
   @ stdin:1
in expression starting at /Users/theabhirath/.julia/packages/MLUtils/2lblx/src/Datasets/datasets.jl:1
in expression starting at /Users/theabhirath/.julia/packages/MLUtils/2lblx/src/Datasets/Datasets.jl:1
in expression starting at /Users/theabhirath/.julia/packages/MLUtils/2lblx/src/MLUtils.jl:1
in expression starting at stdin:1
Stacktrace:
 [1] pkgerror(msg::String)
   @ Pkg.Types ~/julia/usr/share/julia/stdlib/v1.8/Pkg/src/Types.jl:67
 [2] precompile(ctx::Pkg.Types.Context, pkgs::Vector{String}; internal_call::Bool, strict::Bool, warn_loaded::Bool, already_instantiated::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ Pkg.API ~/julia/usr/share/julia/stdlib/v1.8/Pkg/src/API.jl:1405
 [3] precompile
   @ ~/julia/usr/share/julia/stdlib/v1.8/Pkg/src/API.jl:1040 [inlined]
 [4] #precompile#225
   @ ~/julia/usr/share/julia/stdlib/v1.8/Pkg/src/API.jl:1037 [inlined]
 [5] precompile (repeats 2 times)
   @ ~/julia/usr/share/julia/stdlib/v1.8/Pkg/src/API.jl:1037 [inlined]
 [6] top-level scope
   @ REPL[3]:1
```

Reply:
> weird, cannot reproduce this error (which I also don't understand, what's the problem with MLUtisl.Datasets?). You just add the package to Metalhead environment?

Yeah exactly...ironically, outside of the environment adding seems to be working just fine. Is the datasets module conflicting with something else by any chance?

> those timings are crazy, is that some serious precompilation problem for Zygote? what happens if you include("test/runtests.jl") twice?

I have set it to run, will update once I get the result

Reply:
<img width="1133" alt="Screenshot 2022-02-10 at 3 59 08 PM" src="https://user-images.githubusercontent.com/74202102/153388647-4d6fee62-d83a-4f94-9771-a834d3018b45.png">
First and second runs. There's definitely serious first-compile issues here...the ViT `gradtest` is skipped for now until I resolve the MLUtils issue

Reply:
that's crazy, better open a tracking issue.

I'll check the MLUtils issue

Reply:
There is https://github.com/FluxML/Zygote.jl/issues/1119 . Flux master also has https://github.com/FluxML/Flux.jl/pull/1809 which may help some cases, and allows `Chain([Layer(), ...])` which helps some others.

Reply:
The Flux changes will be 0.13 I think.

For the -O1 story, https://github.com/FluxML/Zygote.jl/pull/1147 saw some huge slowdowns elsewhere.

Reply:
https://github.com/JuliaML/MLUtils.jl/issues/48#issue-1130216588 should solve the MLUtils problem